src/components/ContactCard/ContactFieldInput.jsx

```diff
@@ -0,0 +1,93 @@
+import React from "react";
+import PropTypes from "prop-types";
+import { Field } from "react-final-form";
+import classnames from "classnames";
+
+const getInputComponent = inputType =>
+  inputType === "textarea" ? "textarea" : "input";
+
+class ContactFieldInput extends React.Component {
+  state = {
+    renderLabel: false,
+    hasFocus: false
+  };
+
+  onFocus = () => {
+    this.setState({
+      renderLabel: this.props.withLabel,
+      hasFocus: true
+    });
+  };
+
+  onMainInputBlur = e => {
+    this.setState({
+      renderLabel: e.target.value && this.props.withLabel,
+      hasFocus: false
+    });
+  };
+
+  onSecondaryInputBlur = () => {
+    this.setState({
+      hasFocus: false
+    });
+  };
+
+  render() {
+    const {
+      name,
+      type,
+      placeholder,
+      required,
+      withLabel,
+      labelPlaceholder
+    } = this.props;
+    const { renderLabel, hasFocus } = this.state;
+
+    return (
+      <div
+        className={classnames("contact-form__input-wrapper", {
+          "contact-form__input-wrapper--focused": hasFocus
+        })}
+      >
+        <Field
+          name={name}
+          type={type}
+          placeholder={placeholder}
+          required={required}
+          onFocus={this.onFocus}
+          onBlur={this.onMainInputBlur}
+          component={getInputComponent(type)}
+          className="contact-form__input"
+        />
+        {withLabel &&
+          renderLabel && (
+            <Field
+              name={`${name}Label`}
+              type="text"
+              component="input"
+              className="contact-form__label-input"
+              placeholder={labelPlaceholder}
+              onFocus={this.onFocus}
+              onBlur={this.onSecondaryInputBlur}
+            />
+          )}
+      </div>
+    );
+  }
+}
```

Comment:
Regarding the `render` function, I get what you are trying to do. This component named `ContactFieldInput` is not responsible for any input nor any field, but it render a `<Field />` component and another conditional `<Field />` component.
This is not straight forward to understand if the second field will be displayed or not, first because it lays on two boolean `withLabel && renderLabel`, second because it takes a lot of props and distribute those props into two unspecified `<Fields />`.

I recommend:
- to functionally name both underlying components with a semantic name
- to factorize under a named function both boolean with a name that is not `shouldRenderSecondaryField` of course

Those recommendations are not mandatory, just a matter of what could be done to help developers understand what is going on.

Reply:
I'm leaving this as it is for now, but I agree with your sentiment.

To clarify, some of the fields must be able to split themselves into 2 fields, one for the actual value, and one to attach a label to the value.
I'm not sure it would help to create standalone components for them, as they are just a `<Field />`. But the amount of props the parent component receives, and the similar names (`placeholder`, `labelPlaceholder`, etc), are not helping.

I think when we work on editing contacts, we'll have another shot at making this better.

Reply:
I am not sure about the "I agree but won't do it" :-)  
If you agree, do it!

Renaming things may not be very hard, and if it's needed to achieve a good quality and you both agree on it, we should certainly do it and not leave it for tomorow